### DISCUSSION TITLE
Allow member replacement with the replace keyword

### DISCUSSION THREAD

**Discussion by koenbeuk** [dotnet/csharplang#5123] [Language Ideas]
### Updates

1. Uses a new `replaced` keyword rather than a `new` keyword
2. Added alternative proposals

### Summary

Consider allowing the `replace` keyword to be used to replace an existing member on a type, e.g.:
```csharp
class C {
     public static int Foo() => 1;
     replace public static int Foo() => 2;
}

C.Foo(); // returns 2
```
The compiler removes the first implementation of Foo and only keeps the latter.

### Motivation

This allows for some interesting opportunities: 

#### Separation of implementations for defined constants, e.g.:
```csharp
partial class C {
    public void Foo() { 
       // Some default implementation
    }
}

#if NETCOREAPP
public partial class C { 
    replace public void Foo() { 
        // Optimized
    }
}
#endif
``` 

The added advantage here would be that we can shelve away highly optimized implementations in separate files and replace specific members without having to clutter the default implementation's codebase.

#### Source generation with additive mutations:
```csharp
partial class C {
    [Synchronized]
    public void Foo() {
         System.Threading.Thread.Sleep(1000);
    }
}

// A source generator could generate the following:
partial class C {
    [Synchronized]
    replace public void Foo() { 
         lock (this) {
              System.Threading.Thread.Sleep(1000);
         }
    }
}
```
Since the source generator has access to the original method declaration, we can copy it in as shown above- or modify it's implementation which can allow for things like `[AutoConfigureAwait]` which could apply `ConfigureAwait(false)` automatically to each awaitable. In additional, this would allow for:
```csharp
[AutoPropertyChanged]
public string Name { get; protected set; }

// generates:
[AutoPropertyChanged]
private string _name;
public string Name { get => _name; protected set { _name = value; RaisePropertyChanged("Name"); } }
```

#### Customizing generated code.
```csharp
// Generated class
public partial class FooFactory { 
     string GetNextId() => Guid.NewGuid.ToString();

     public Foo CreateFoo() { 
         return Foo { Id = GetNextId() };
     }
}
```

And a user can then replace functionality as seem fit:
```csharp
// User code
public partial class FooFactory {
    replace string GetNextId() => SomeIdGenerator();
}
```
### Design
Members can be replaced within the same class by a member with a matching signature with the `replace` keyword applied. At most 2 declarations of the same member can be present within a type and at most one has to have the `replace` keyword applied. When a member is replaced, the compiler will only include the replaced version of that member.

### Drawbacks
This proposal has overlap with partial methods. Partial methods are more restricted as they do not concern with properties and are not able to provide a default body.

### Alternative proposals
#1070 deals with a similar keyword yet has a different meaning. Instead of completely replacing an existing implementation with a new implementation as in this proposal, 107 keeps a reference to the replaced implementation which can then still be called by the replacing implementation through a new keyword (`original()`); This allows for multiple replacers over a single method where this proposal is limited to a single replacement.

#3772 offers a similar proposal as #1070 while restricting to additive only mutations.

**Comment by HaloFour** [comment]
This is the same as [`replace`/`original`](https://github.com/dotnet/csharplang/issues/107) but with a different keyword.  It has the same problems with ordering.  The team has also expressed disinterest with using source generators for AOP scenarios.

**Reply by koenbeuk** [reply]
I will read more into that proposal however there seem to be some differences that are noteworthy for now:

1. This proposal does not introduce a new word or a new concept. #107 would introduce both a new `supersede` keyword which arguably is better than `new` as it as well as a new `superseded` keyword that can be used to invoke member that was superseded. 
2. And because of not introducing a `superseded` keyword, This proposal does not introduce a new ordering system that needs to be implemented. A member can only be superseded/replaced once meaning that there is no concept of ordering which greatly simplifies the impact. In general this proposal is about making minimal changes

**Reply by HaloFour** [reply]
The keywords themselves aren't important, it's the concepts they convey.  Adding a new contextual keyword is not expensive and would likely be preferred, especially given that `new` in a method signature already means something.  You can argue that this is a form of "shadowing", but that does imply that the original member is still there and still accessible whereas you're suggesting that it's completely removed by the compiler.  You also point this out by virtue of requiring doubling up `new new` in order to achieve both, which to me is reason enough to avoid trying to overload the existing keyword.

**Reply by koenbeuk** [reply]
Thanks for the input, I indeed concur that `new` is not the best keyword for this. I've updated the proposal to use a `replace` keyword instead as is used in #107.

**Reply by koenbeuk** [reply]
For clarification: I consider this proposal different from #107 as this proposal does not deal with shadowing existing members or with the complexities that come with that such as ordering. This proposal tries to be as close in line with what we have today with source generators and partial methods. Only ever one instance of a member can exist, just like with partial methods and source generated members. 

This indeed means that multiple source generators are not allowed to replace the same member. This will cause an error. Also, source generators will never receive the generated replacement. They will always receive the original which is in line with how source generators work today.

Scenarios like having multiple decorators applied on a member are in this way still possible for as long as the source generator knows how to deal with things:

```csharp
[Decorated(typeof(Measured), typeof(Logged), typeof(Synchronized))]
```
